### PR TITLE
Inset xranges

### PR DESCRIPTION
**Pull request by vext01**
This adds support for xranges to insets.

### CODE REVIEW

**Comment by snim2** [comment]
Fixes #356

**Comment by vext01** [comment]
Try this.

File: bin/plot_krun_results [hunk 1]
```diff
@@ -667,24 +669,32 @@ class ProcessExecChart(object):
...
+                # If all the changepoints occur at the start of the process exec,
+                # include them all and a few extra executions.
+                if self.changepoints:
+                    if (self.changepoints[self.x_bounds[0]:self.x_bounds[1]] and
+                          self.changepoints[self.x_bounds[0]:self.x_bounds[1]][-1] <= int(one_pc * 5.0)):
+                        inset_xlimit = (0, (self.changepoints[self.x_bounds[0]:self.x_bounds[1]][-1]
+                                            + int(one_pc * 2.0)))
+            else:
+                inset_xlimit = self.inset_xlimit
+                # Clamp the extent of the inset.
```
Comment (snim2) [inline bin/plot_krun_results]: Surely we only need to clamp if the user has *not* specified the inset xlimits?

File: bin/plot_krun_results [hunk 2]
```diff
@@ -698,24 +708,13 @@ class ProcessExecChart(object):
             self._plot_outliers(inset, large_markers=False)
             self._plot_changepoints(inset, large_markers=False)
             pyplot.draw()  # Force xticklabels to be drawn.
-            if self.x_bounds[0] > 0:
```
Comment (snim2) [inline bin/plot_krun_results:701]: Let's do some testing to see whether we need this...

File: bin/plot_krun_results [hunk 3]
```diff
@@ -1297,6 +1304,12 @@ def create_cli_parser():
                         default=None,
                         help='Only plot the core cycles for specified cores'
                               '(comma separated). e.g: 0,1,2,3')
+    parser.add_argument('--inset-xlimits', '-X',
+                        action='store',
+                        dest='inset_xlimits',
+                        default=None,
+                        type=str,
+                        help="Similar to --xlimits, but for the inset plots.")
```
Comment (snim2) [inline bin/plot_krun_results]: Single quotes.

File: bin/plot_krun_results [hunk 4]
```diff
@@ -1356,6 +1369,30 @@ if __name__ == '__main__':
...
+        except ValueError:
+            fatal_error('invalid --xlimits pair')
+
+        if len(options.xlimits) != 2:
+            fatal_error('wrong number of --xlimits elements')
+
+    if options.inset_xlimits:
+        try:
+            options.inset_xlimits = \
+                [int(x) for x in options.inset_xlimits.split(",")]
```
Comment (snim2) [inline bin/plot_krun_results]: Single quotes.

File: bin/plot_krun_results [hunk 5]
```diff
@@ -1356,6 +1369,30 @@ if __name__ == '__main__':
     else:
         core_cycles = None
 
+    if options.xlimits:
+        try:
+            options.xlimits = [int(x) for x in options.xlimits.split(",")]
```
Comment (snim2) [inline bin/plot_krun_results]: Single quotes.

**Comment by snim2** [comment]
Ok, some minor comments and I think we need some more testing to see whether munging the inset `xticklabels` is really necessary.

File: bin/plot_krun_results (hunk 2 above)
Comment (vext01) [reply bin/plot_krun_results:701]: [yay.pdf](https://github.com/softdevteam/warmup_experiment/files/868320/yay.pdf)

Here's some 0.7 data. Do you see any weirdness with the ticks?

I'll plot some more with less regular ticks...

Comment (vext01) [reply bin/plot_krun_results:701]: More examples with wierd xticks.
[yay2.pdf](https://github.com/softdevteam/warmup_experiment/files/868388/yay2.pdf)

Comment (vext01) [reply bin/plot_krun_results:701]: With no arguments:
[noargs.pdf](https://github.com/softdevteam/warmup_experiment/files/868477/noargs.pdf)

**Comment by snim2** [comment]
V0.7 OpenBSD data (no CLI arguments) seems fine.

File: bin/plot_krun_results (hunk 2 above)
Comment (snim2) [reply bin/plot_krun_results:701]: On yay2.pdf, the x-ticklabels on the wallclock data chart have become floats, and the xticklabels on the inset don't seem to include the first and last datum.

Comment (snim2) [reply bin/plot_krun_results:701]: OK, so it seems that if we give the script more sensible --xlimits then the xticklabels are integers, so I think we can leave that issue

**Comment by snim2** [comment]
OK, let's squash!

**Comment by vext01** [comment]
Kersplat

**Review by snim2 — APPROVED** [review]

**Pull request merged**